course5/homework10.2/linked_list.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <iterator>
+#include <memory>
+
+namespace jds {
+
+template <typename T>
+class LinkedList
+{
+private:
+	struct Node
+	{
+		T value;
+		std::unique_ptr<Node> next{nullptr};
+	};
+
+	std::unique_ptr<Node> head{nullptr};
+public:
+	class iterator: public std::forward_iterator_tag {
+	public:
+		iterator() = default;
+		iterator(Node * initial): current(initial) { }
+		iterator(iterator const & that) = default;
+		iterator & operator=(iterator const & that) { this->current = that.current; return *this; }
+		iterator & operator++() {
+			current = current->next.get();
+			return *this;
+		}
+		iterator operator++(int) {
+			iterator old = *this;
+			current = current->next.get();
+			return old;
+		}
+		T & operator*() const { return current->value; }
+		bool operator==(iterator const & that) const { return this->current == that.current; }
+		bool operator!=(iterator const & that) const { return this->current != that.current; }
+	private:
+		LinkedList<T>::Node * current{nullptr};
+	};
+
+	class const_iterator: public std::forward_iterator_tag {
+	public:
+		const_iterator() = default;
+		const_iterator(const_iterator const & that) = default;
+		const_iterator & operator=(const_iterator const & that) { this->current = that.current; return *this; }
+		const_iterator(Node const * initial): current(initial) { }
+		const_iterator & operator++() {
+			current = current->next.get();
+			return *this;
+		}
+		const_iterator operator++(int) {
+			const_iterator old = *this;
+			current = current->next.get();
+			return old;
+		}
+		T const & operator*() const { return current->value; }
+		bool operator==(const_iterator const & that) const { return this->current == that.current; }
+		bool operator!=(const_iterator const & that) const { return this->current != that.current; }
+	private:
+		Node const * current{nullptr};
+	};
+
+	LinkedList() = default;
+	LinkedList(LinkedList const & that): LinkedList(that.begin(), that.end()) { }
+	LinkedList(LinkedList && that): head(std::move(that.head)) { }
+
+	template <class InputIt>
+	LinkedList(InputIt const & first, InputIt const & last)
+	{
+		std::unique_ptr<Node> * current{&head};
```

Comment:
Why are you using this curly-bracket initialisation everywhere? Makes code harder to read.